File: mvvmsmart/src/main/java/com/wzq/mvvmsmart/http/download/ProgressCallBack.java
package com.wzq.mvvmsmart.http.download;

import android.util.Log;

import com.jeremyliao.liveeventbus.LiveEventBus;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;

import androidx.annotation.Nullable;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.Observer;
import okhttp3.ResponseBody;


public abstract class ProgressCallBack<T> {

    private String destFileDir; // 本地文件存放路径
    private String destFileName; // 文件名

    public ProgressCallBack(LifecycleOwner lifecycleOwner, String destFileDir, String destFileName) {
        this.destFileDir = destFileDir;
        this.destFileName = destFileName;
        subscribeLoadProgress(lifecycleOwner);
    }

    public abstract void onSuccess(T t);

    public abstract void progress(long progress, long total);

    public void onStart() {
    }

    public void onCompleted() {
    }

    public abstract void onError(Throwable e);

    public void saveFile(ResponseBody body) {
        InputStream is = null;
        byte[] buf = new byte[2048];
        int len;
        FileOutputStream fos = null;
        try {
            is = body.byteStream();
            File dir = new File(destFileDir);
            if (!dir.exists()) {
                dir.mkdirs();
            }
            File file = new File(dir, destFileName);
            fos = new FileOutputStream(file);
            while ((len = is.read(buf)) != -1) {
                fos.write(buf, 0, len);
            }
            fos.flush();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                if (is != null) is.close();
                if (fos != null) fos.close();
            } catch (IOException e) {
                Log.e("saveFile", e.getMessage());
            }
        }
    }

    /**
     * 订阅加载的进度条
     *
     * @param lifecycleOwner
     */
    public void subscribeLoadProgress(LifecycleOwner lifecycleOwner) {
        LiveEventBus
                .get("key_name", DownLoadStateBean.class)
                .observe(lifecycleOwner, new Observer<DownLoadStateBean>() {
                    @Override
                    public void onChanged(@Nullable DownLoadStateBean progressLoadBean) {
                        progress(progressLoadBean.getBytesLoaded(), progressLoadBean.getTotal());
                    }
                });
    }


}

File: mvvmsmart/src/main/java/com/wzq/mvvmsmart/net/download/ProgressCallBack.kt
package com.wzq.mvvmsmart.net.download

import android.util.Log
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.Observer
import com.jeremyliao.liveeventbus.LiveEventBus
import okhttp3.ResponseBody
import java.io.*

abstract class ProgressCallBack<T>(lifecycleOwner: LifecycleOwner?, // 本地文件存放路径
                                   private val destFileDir: String, // 文件名
                                   private val destFileName: String) {

    abstract fun onSuccess(t: Any?)
    abstract fun progress(progress: Long, total: Long)
    open fun onStart() {}
    open fun onCompleted() {}
    abstract fun onError(e: Throwable?)
    init {
        subscribeLoadProgress(lifecycleOwner)
    }
    fun saveFile(body: ResponseBody) {
        var `is`: InputStream? = null
        val buf = ByteArray(2048)
        var len: Int
        var fos: FileOutputStream? = null
        try {
            `is` = body.byteStream()
            val dir = File(destFileDir)
            if (!dir.exists()) {
                dir.mkdirs()
            }
            val file = File(dir, destFileName)
            fos = FileOutputStream(file)
            while (`is`.read(buf).also { len = it } != -1) {
                fos.write(buf, 0, len)
            }
            fos.flush()
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        } finally {
            try {
                `is`?.close()
                fos?.close()
            } catch (e: IOException) {
                Log.e("saveFile", e.message)
            }
        }
    }

    /**
     * 订阅加载的进度条
     * @param lifecycleOwner
     */
    fun subscribeLoadProgress(lifecycleOwner: LifecycleOwner?) {
        LiveEventBus
                .get("key_name", DownLoadStateBean::class.java)
                .observe(lifecycleOwner!!, Observer { progressLoadBean -> progress(progressLoadBean.bytesLoaded, progressLoadBean.total) })
    }


}